Sim/SimPapas/src/PapasImportParticlesTool.cpp

```diff
@@ -35,31 +36,35 @@ StatusCode PapasImportParticlesTool::clear() {
 
 StatusCode PapasImportParticlesTool::finalize() { return GaudiTool::finalize(); }
 
-StatusCode PapasImportParticlesTool::run(papas::Event& pevent) {
-
+StatusCode PapasImportParticlesTool::run(papas::Event& pevent, std::unordered_map<papas::Identifier, int>& links, std::shared_ptr<papas::Detector> spDetector) {
   const fcc::MCParticleCollection* ptcs = m_iGenpHandle.get();
-  auto detector = papas::CMS();  // todo consider making this be passed as part of the papas tool interface
-
   int countp = 0;
 
-  // First Sort fcc MCparticles in order of decreasing energy to match Python
+  // We need to know the sorted order of fcc MCparticles by decreasing energy to match Python
   // Sort is required when run in PDebug mode. TODO: make this configurable .
-  std::list<fcc::ConstMCParticle> sortPtcs;
+  std::vector<fcc::ConstMCParticle> vecPtcs;
+  std::list<int> orderPtcs; // will contain the GenParticle index for the sorted Genparticles
+  int n=0;
   for (const auto& p : *ptcs) {
-    sortPtcs.push_back(p);
+    orderPtcs.push_back(n);
+    n++;
+    vecPtcs.push_back(p);
   }
   TLorentzVector tlv;
   TLorentzVector tlv2;
-  sortPtcs.sort([&tlv, &tlv2](const fcc::ConstMCParticle& a, const fcc::ConstMCParticle& b) {
-    auto p4 = a.p4();
+  
+  //order the list of the indices so as to correspond to GenParticles in decreasing order of energy
+  orderPtcs.sort([&tlv, &tlv2, &vecPtcs](int a, int b) {
+    auto p4 = vecPtcs[a].p4();
     tlv.SetXYZM(p4.px, p4.py, p4.pz, p4.mass);
-    p4 = b.p4();
+    p4 = vecPtcs[b].p4();
     tlv2.SetXYZM(p4.px, p4.py, p4.pz, p4.mass);
     return tlv.E() > tlv2.E();
   });
-
+  
   // Now construct the papas Particles (providing they are stable)
```

Comment:
need more explicit comment. What kind of papas Particles are constructed? 
I understand that we are talking about the reconstructed particles from the comment,
but the code seems to indicate we are constructing the simulated particles before papas

Reply:
Improved